src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs

```diff
@@ -1269,26 +1277,115 @@ public static bool IsPatternContext(this SyntaxTree syntaxTree, SyntaxToken left
+            if (leftToken.IsKind(SyntaxKind.OpenParenToken, SyntaxKind.CommaToken) && node.IsKind(SyntaxKind.PositionalPatternClause))
             {
                 return true;
             }
 
             // e is { P: $$
             // e is { ..., P: $$
-            if (leftToken.IsKind(SyntaxKind.ColonToken) && leftToken.Parent.IsKind(SyntaxKind.NameColon) &&
-                leftToken.Parent.IsParentKind(SyntaxKind.Subpattern))
+            if (leftToken.IsKind(SyntaxKind.ColonToken) && node.IsKind(SyntaxKind.NameColon) &&
+                node.IsParentKind(SyntaxKind.Subpattern))
+            {
+                return true;
+            }
+
+#if !CODE_STYLE
+            // e switch { ($$
+            // e switch { ..., ($$
+            if (leftToken.IsKind(SyntaxKind.OpenParenToken) && node.IsParentKind(SyntaxKind.SwitchExpressionArm))
+            {
+                return true;
+            }
+
+            // case ($
+            if (leftToken.IsKind(SyntaxKind.OpenParenToken) && node.IsParentKind(SyntaxKind.SwitchSection))
+            {
+                return true;
+            }
+
+            // e is ((($$
+            // e is { P: ($$
+            if (leftToken.IsKind(SyntaxKind.OpenParenToken) && node.IsKind(SyntaxKind.ConstantPattern))
             {
                 return true;
             }
 
+            // if (e is { P: (1 or $$
+            if ((leftToken.IsKind(SyntaxKind.AndKeyword) || leftToken.IsKind(SyntaxKind.OrKeyword)) &&
+                node.IsParentKind(SyntaxKind.ParenthesizedPattern))
+            {
+                return true;
+            }
+
+            if (node.IsParentKind(SyntaxKind.IsPatternExpression))
+            {
+                // e is 1 and $$
+                if (node.IsKind(SyntaxKind.AndPattern))
+                {
+                    return true;
+                }
+
+                // e is 1 or $$
+                if (node.IsKind(SyntaxKind.OrPattern))
+                {
+                    return true;
+                }
+
+                // e is not $$
+                if (node.IsKind(SyntaxKind.NotPattern))
+                {
+                    return true;
+                }
+            }
+#endif
+
+            return false;
+        }
+
+        public static bool IsAfterPatternContext(this SyntaxTree syntaxTree, SyntaxToken leftToken, int position)
+        {
+#if !CODE_STYLE
+            leftToken = leftToken.GetPreviousTokenIfTouchingWord(position);
+
+            // e is 1 $$
+            if (leftToken.Parent.IsParentKind(SyntaxKind.ConstantPattern))
+            {
+                return true;
+            }
+
+            // e is int $$
+            if (leftToken.Parent.IsParentKind(SyntaxKind.DeclarationPattern))
+            {
+                return true;
+            }
+
+            // e is >= 0 $$
+            if (leftToken.Parent.IsParentKind(SyntaxKind.RelationalPattern))
```

Comment:
this is not sufficient.  say you have `e is >= a.b.c.d.e.f.g.h`, then `Parent.IsParentKind` won't work.  you need to walk up the rhs to see if you get a pattern, and see if that pattern ends where you currently are.